File: src/y2022/day2.jl
module Day2

export day2

const POINTS = Dict{String, Int8}(
                "win" => 6,
                "draw" => 3,
                "loss" => 0,
                "rock" => 1,
                "paper" => 2,
                "scissor" => 3,
               )

const STRATEGY_GUIDE_1 = Dict{String, Vector{String}}(
                               "A X" => ["draw", "rock"],
                               "A Y" => ["win", "paper"],
                               "A Z" => ["loss", "scissor"],
                               "B X" => ["loss", "rock"],
                               "B Y" => ["draw", "paper"],
                               "B Z" => ["win", "scissor"],
                               "C X" => ["win", "rock"],
                               "C Y" => ["loss", "paper"],
                               "C Z" => ["draw", "scissor"],
                              )

const STRATEGY_GUIDE_2 = Dict{String, Vector{String}}(
                               "A X" => ["loss", "scissor"],
                               "A Y" => ["draw", "rock"],
                               "A Z" => ["win", "paper"],
                               "B X" => ["loss", "rock"],
                               "B Y" => ["draw", "paper"],
                               "B Z" => ["win", "scissor"],
                               "C X" => ["loss", "paper"],
                               "C Y" => ["draw", "scissor"],
                               "C Z" => ["win", "rock"],
                              )


function calc_round_points_1(strategy::String)::Int8
    return sum([POINTS[result] for result in STRATEGY_GUIDE_1[strategy]])
end

function calc_round_points_2(strategy::String)::Int8
    return sum([POINTS[result] for result in STRATEGY_GUIDE_2[strategy]])
end


function day2()
  # Import data
  data = open("./2022/day2.txt", "r");
  strategy_data = chomp(read(data, String))
  close(data)

  strategy_list::Vector{String} = [
                 game_strategy
                 for game_strategy in split(strategy_data, "\n")
                ]

  #################################
#
  round_points::Vector{Int8} = map(calc_round_points_1, strategy_list)
  total_points = sum(round_points)

  results = [total_points]

  ###########################

  round_points = map(calc_round_points_2, strategy_list)
  total_points = sum(round_points)

  push!(results, total_points)
  println("Day 2 Results - ", results)

end

end

File: src/y2022/day2.rs
use std::collections::HashMap;
use std::fs;


pub fn day2() -> [i32; 2] {
    let result = [p1(), p2()];
    return result;
}


fn p1() -> i32 {
    let strategy = HashMap::from([
        ("X", 1), //rock
        ("Y", 2), //paper
        ("Z", 3), //scissor
        ("A", 1), //rock
        ("B", 2), //paper
        ("C", 3), //scissor
    ]);

    let data = load_data();

    let mut points = 0;

    for line in data {
        let (opponent, me) = line.split_once(" ").expect("No whitespace");
        let result = strategy.get(opponent).expect("not found") - strategy.get(me).expect("not found");
        match result {
            -1 | 2 => points += strategy.get(me).expect("not found") + 6,
            0 => points += strategy.get(me).expect("not found") + 3,
            1 | -2 => points += strategy.get(me).expect("not found") + 0,
            _ => panic!("shouldnt happen")
        }
    }
    return points;
}

fn p2() -> i32 {
    let strategy = HashMap::from([
        ("X", 0), //loss
        ("Y", 3), //draw
        ("Z", 6), //win
        ("A", 1), //rock
        ("B", 2), //paper
        ("C", 3), //scissor
    ]);

    let data = load_data();

    let mut points = 0;

    for line in data {
        let (opponent, result) = line.split_once(" ").expect("No whitespace");
        match result {
            "X" => points += strategy.get(result).expect("not found") + ((((strategy.get(opponent).unwrap() - 1) + 2) % 3) + 1),
            "Y" => points += strategy.get(result).expect("not found") + strategy.get(opponent).unwrap(),
            "Z" => points += strategy.get(result).expect("not found") + ((((strategy.get(opponent).unwrap() - 1) + 1) % 3) + 1),
            _ => panic!("Should not happen.")
        }

    }
    return points;
}

fn load_data() -> Vec<String> {
    let data = fs::read_to_string("./2022/day2.txt")
        .expect("Couldn't read data :(");
    let lines: Vec<&str> = data.lines().collect();
    let owned_strings: Vec<String> = lines.iter().map(|&line| String::from(line)).collect();
    return owned_strings;
}
